components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1269,7 +1270,139 @@ where
             SignificantMsg::RaftlogFetched { context, res } => {
                 self.on_raft_log_fetched(context, res);
             }
+            SignificantMsg::EnterForceLeaderState { failed_stores } => {
+                self.on_enter_pre_force_leader(failed_stores)
+            }
+            SignificantMsg::ExitForceLeaderState => self.on_exit_force_leader(),
+        }
+    }
+
+    fn on_enter_pre_force_leader(&mut self, failed_stores: HashSet<u64>) {
+        if self.fsm.peer.force_leader.is_some() {
+            return;
+        }
+
+        let region = self.region();
+        let expected_alive_voter = self
```

Comment:
So, why not use `self.region().get_peers().iter().filter()`?

Reply:
need to decide which are voters and which aren't, better leave raft to do this

Reply:
They should match each other exactly. Using `region` seems more readable.

Reply:
yes, they are the same. But I want to reuse the logic of raft to decide the voters considering the joint consensus.